File: JSONParser/JSONParser.m
//
//  JSONParser.m
//  JSONParser
//
//  Created by Cory Kilger on 4/20/15.
//  Copyright (c) 2015 Cory Kilger. All rights reserved.
//

#import "JSONParser.h"
#import "json_parser.h"

NSString * JSONParserErrorDomain = @"JSONParserErrorDomain";

@implementation JSONParser

void startObject(void *userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parserDidStartObject:parser];
}

void endObject(void *userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parserDidEndObject:parser];
}

void startArray(void *userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parserDidStartArray:parser];
}

void endArray(void *userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parserDidEndArray:parser];
}

void foundString(const char * string, uint64_t length, void * userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parser:parser foundString:CFBridgingRelease(CFStringCreateWithBytes(NULL, (const UInt8 *)string, (CFIndex)length, kCFStringEncodingUTF8, false))];
}

void foundNumber(double number, void * userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parser:parser foundNumber:CFBridgingRelease(CFNumberCreate(NULL, kCFNumberDoubleType, &number))];
}

void foundTrue(void *userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parserFoundTrue:parser];
}

void foundFalse(void *userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parserFoundFalse:parser];
}

void foundNull(void *userInfo) {
    JSONParser * parser = (__bridge JSONParser *)(userInfo);
    [parser->_delegate parserFoundNull:parser];
}

- (BOOL)parseData:(NSData *)data error:(NSError **)error {
    json_parser parser;
    parser.userInfo = (__bridge void *)(self);
    parser.startObject = startObject;
    parser.endObject = endObject;
    parser.startArray = startArray;
    parser.endArray = endArray;
    parser.foundString = foundString;
    parser.foundNumber = foundNumber;
    parser.foundTrue = foundTrue;
    parser.foundFalse = foundFalse;
    parser.foundNull = foundNull;
    
    uint8_t parseError;
    uint64_t length;
    size_t size = data.length;
    char * buffer = malloc(size);
    parseJSON(data.bytes, &length, &parseError, &parser, buffer, size);
    free(buffer);
    
    if (error) {
        if (parseError && error) {
            NSDictionary * userInfo = @{NSLocalizedDescriptionKey: NSLocalizedString(@"JSON is not in the correct format", @"Error message when failing to parse JSON")};
            *error = [NSError errorWithDomain:JSONParserErrorDomain code:1000 userInfo:userInfo];
        } else {
            *error = nil;
        }
    }
    
    return (parseError == 0);
}

@end

File: JSONParser/JSONParser.swift
//
//  JSONParser.swift
//  JSONParser
//
//  Created by Cory Kilger on 4/20/15.
//  Copyright (c) 2015 Cory Kilger. All rights reserved.
//

import Foundation

@objc
protocol JSONParserDelegate: class {
    optional func parserDidStartObject(parser: JSONParser)
    optional func parserDidEndObject(parser: JSONParser)
    optional func parserDidStartArray(parser: JSONParser)
    optional func parserDidEndArray(parser: JSONParser)
    optional func parser(parser: JSONParser, foundString: String)
    optional func parser(parser: JSONParser, foundNumber: NSNumber)
    optional func parserFoundTrue(parser: JSONParser)
    optional func parserFoundFalse(parser: JSONParser)
    optional func parserFoundNull(parser: JSONParser)
}

class JSONParser: NSObject {
    weak var delegate: JSONParserDelegate?
    
    func parseData(data: NSData) {
        var parser = json_parser()
        
        let startObject = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        startObject.initialize { (parser) -> Void in
            
        }
        parser.startObject = CFunctionPointer(COpaquePointer(startObject))
        
        
        let endObject = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        endObject.initialize { (parser) -> Void in
            
        }
        parser.endObject = CFunctionPointer(COpaquePointer(endObject))
        
        let startArray = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        startArray.initialize { (parser) -> Void in
            
        }
        parser.startArray = CFunctionPointer(COpaquePointer(startArray))
        
        let endArray = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        endArray.initialize { (parser) -> Void in
            
        }
        parser.endArray = CFunctionPointer(COpaquePointer(endArray))
        
        let foundString = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        foundString.initialize { (parser) -> Void in
            
        }
        parser.foundString = CFunctionPointer(COpaquePointer(foundString))
        
        let foundNumber = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        foundNumber.initialize { (parser) -> Void in
            
        }
        parser.foundNumber = CFunctionPointer(COpaquePointer(foundNumber))
        
        let foundTrue = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        foundTrue.initialize { (parser) -> Void in
            
        }
        parser.foundTrue = CFunctionPointer(COpaquePointer(foundTrue))
        
        let foundFalse = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        foundFalse.initialize { (parser) -> Void in
            
        }
        parser.foundFalse = CFunctionPointer(COpaquePointer(foundFalse))
        
        let foundNull = UnsafeMutablePointer<(userInfo: UnsafeMutablePointer<Void>) -> Void>.alloc(1)
        foundNull.initialize { (parser) -> Void in
            
        }
        parser.foundNull = CFunctionPointer(COpaquePointer(foundNull))
    }
}
